File: benchmark_performance.py
"""
Performance benchmarking script
Compares inefficient vs optimized implementations
"""
import time
import random
import json
import tempfile
from data_processor import (
    process_user_data as process_old,
    find_duplicates as find_dup_old,
    calculate_statistics as calc_stats_old,
    filter_and_transform as filter_old,
    DataCache as CacheOld,
    process_large_dataset as process_dataset_old,
    generate_report as gen_report_old
)
from data_processor_optimized import (
    process_user_data as process_new,
    find_duplicates as find_dup_new,
    calculate_statistics as calc_stats_new,
    filter_and_transform as filter_new,
    DataCache as CacheNew,
    process_large_dataset as process_dataset_new,
    generate_report as gen_report_new
)


def benchmark(func, *args, iterations=100):
    """Run benchmark and return average time"""
    times = []
    for _ in range(iterations):
        start = time.perf_counter()
        func(*args)
        end = time.perf_counter()
        times.append(end - start)
    return sum(times) / len(times)


def generate_test_data():
    """Generate test data for benchmarks"""
    # User data
    users = [
        {'id': i, 'name': f'User{i}', 'active': i % 2 == 0, 'value': i * 10}
        for i in range(1000)
    ]
    
    # List with duplicates
    items = [random.randint(0, 100) for _ in range(1000)]
    
    # Numbers for statistics
    numbers = [random.random() * 100 for _ in range(1000)]
    
    # Data for filtering
    data = list(range(1000))
    
    # Large dataset
    large_dataset = list(range(10000))
    
    # Records for report
    records = [
        {'id': i, 'name': f'Record{i}', 'value': i * 5}
        for i in range(100)
    ]
    
    return {
        'users': users,
        'items': items,
        'numbers': numbers,
        'data': data,
        'large_dataset': large_dataset,
        'records': records
    }


def run_benchmarks():
    """Run all benchmarks and display results"""
    print("=" * 80)
    print("PERFORMANCE BENCHMARK RESULTS")
    print("=" * 80)
    print()
    
    test_data = generate_test_data()
    
    # Benchmark 1: Process user data
    print("1. Process User Data (filter + transform)")
    print("-" * 80)
    old_time = benchmark(process_old, test_data['users'])
    new_time = benchmark(process_new, test_data['users'])
    improvement = ((old_time - new_time) / old_time) * 100
    print(f"   Inefficient: {old_time*1000:.4f} ms")
    print(f"   Optimized:   {new_time*1000:.4f} ms")
    print(f"   Improvement: {improvement:.1f}% faster")
    print()
    
    # Benchmark 2: Find duplicates
    print("2. Find Duplicates")
    print("-" * 80)
    old_time = benchmark(find_dup_old, test_data['items'])
    new_time = benchmark(find_dup_new, test_data['items'])
    improvement = ((old_time - new_time) / old_time) * 100
    print(f"   Inefficient: {old_time*1000:.4f} ms")
    print(f"   Optimized:   {new_time*1000:.4f} ms")
    print(f"   Improvement: {improvement:.1f}% faster")
    print()
    
    # Benchmark 3: Calculate statistics
    print("3. Calculate Statistics")
    print("-" * 80)
    old_time = benchmark(calc_stats_old, test_data['numbers'])
    new_time = benchmark(calc_stats_new, test_data['numbers'])
    improvement = ((old_time - new_time) / old_time) * 100
    print(f"   Inefficient: {old_time*1000:.4f} ms")
    print(f"   Optimized:   {new_time*1000:.4f} ms")
    print(f"   Improvement: {improvement:.1f}% faster")
    print()
    
    # Benchmark 4: Filter and transform
    print("4. Filter and Transform")
    print("-" * 80)
    old_time = benchmark(filter_old, test_data['data'], 500)
    new_time = benchmark(filter_new, test_data['data'], 500)
    improvement = ((old_time - new_time) / old_time) * 100
    print(f"   Inefficient: {old_time*1000:.4f} ms")
    print(f"   Optimized:   {new_time*1000:.4f} ms")
    print(f"   Improvement: {improvement:.1f}% faster")
    print()
    
    # Benchmark 5: Cache operations
    print("5. Cache Operations (1000 get/set operations)")
    print("-" * 80)
    
    def test_cache(CacheClass):
        cache = CacheClass()
        for i in range(1000):
            cache.set(f'key{i}', f'value{i}')
        for i in range(1000):
            cache.get(f'key{i % 1000}')
    
    old_time = benchmark(test_cache, CacheOld, iterations=10)
    new_time = benchmark(test_cache, CacheNew, iterations=10)
    improvement = ((old_time - new_time) / old_time) * 100
    print(f"   Inefficient: {old_time*1000:.4f} ms")
    print(f"   Optimized:   {new_time*1000:.4f} ms")
    print(f"   Improvement: {improvement:.1f}% faster")
    print()
    
    # Benchmark 6: Process large dataset
    print("6. Process Large Dataset")
    print("-" * 80)
    old_time = benchmark(process_dataset_old, test_data['large_dataset'], iterations=10)
    new_time = benchmark(process_dataset_new, test_data['large_dataset'], iterations=10)
    improvement = ((old_time - new_time) / old_time) * 100
    print(f"   Inefficient: {old_time*1000:.4f} ms")
    print(f"   Optimized:   {new_time*1000:.4f} ms")
    print(f"   Improvement: {improvement:.1f}% faster")
    print()
    
    # Benchmark 7: Generate report
    print("7. Generate Report")
    print("-" * 80)
    old_time = benchmark(gen_report_old, test_data['records'])
    new_time = benchmark(gen_report_new, test_data['records'])
    improvement = ((old_time - new_time) / old_time) * 100
    print(f"   Inefficient: {old_time*1000:.4f} ms")
    print(f"   Optimized:   {new_time*1000:.4f} ms")
    print(f"   Improvement: {improvement:.1f}% faster")
    print()
    
    print("=" * 80)
    print("SUMMARY")
    print("=" * 80)
    print("All optimizations show measurable performance improvements.")
    print("Key optimization techniques used:")
    print("  - List comprehensions instead of loops")
    print("  - Built-in functions (sum, sorted, Counter)")
    print("  - Dict/Set for O(1) lookups instead of lists")
    print("  - String join instead of concatenation")
    print("  - Single-pass algorithms instead of multiple passes")
    print("=" * 80)


if __name__ == '__main__':
    run_benchmarks()

File: benchmark_performance.js
/**
 * JavaScript Performance Benchmarking
 * Compares inefficient vs optimized implementations
 */

const inefficient = require('./inefficient_code');
const optimized = require('./optimized_code');

/**
 * Benchmark a function
 */
function benchmark(fn, args, iterations = 1000) {
    const times = [];
    
    for (let i = 0; i < iterations; i++) {
        const start = performance.now();
        fn(...args);
        const end = performance.now();
        times.push(end - start);
    }
    
    const avg = times.reduce((a, b) => a + b, 0) / times.length;
    return avg;
}

/**
 * Generate test data
 */
function generateTestData() {
    const items = Array.from({ length: 1000 }, (_, i) => ({
        id: i,
        name: `Item${i}`,
        active: i % 2 === 0,
        value: i * 10
    }));
    
    const duplicateArray = Array.from({ length: 1000 }, () => 
        Math.floor(Math.random() * 100)
    );
    
    const numbers = Array.from({ length: 1000 }, () => 
        Math.floor(Math.random() * 1000)
    );
    
    const data = Array.from({ length: 1000 }, (_, i) => ({
        title: `Title ${i}`,
        description: `Description for item ${i}`,
        value: i
    }));
    
    const emails = Array.from({ length: 1000 }, (_, i) => 
        i % 3 === 0 ? `user${i}@example.com` : `invalid-email-${i}`
    );
    
    return {
        items,
        duplicateArray,
        numbers,
        data,
        emails
    };
}

/**
 * Run all benchmarks
 */
function runBenchmarks() {
    console.log('='.repeat(80));
    console.log('JAVASCRIPT PERFORMANCE BENCHMARK RESULTS');
    console.log('='.repeat(80));
    console.log();
    
    const testData = generateTestData();
    
    // Benchmark 1: Process items
    console.log('1. Process Items (filter + map + sort)');
    console.log('-'.repeat(80));
    const time1Old = benchmark(inefficient.processItems, [testData.items]);
    const time1New = benchmark(optimized.processItems, [testData.items]);
    const improvement1 = ((time1Old - time1New) / time1Old) * 100;
    console.log(`   Inefficient: ${time1Old.toFixed(4)} ms`);
    console.log(`   Optimized:   ${time1New.toFixed(4)} ms`);
    console.log(`   Improvement: ${improvement1.toFixed(1)}% faster`);
    console.log();
    
    // Benchmark 2: Find duplicates
    console.log('2. Find Duplicates');
    console.log('-'.repeat(80));
    const time2Old = benchmark(inefficient.findDuplicates, [testData.duplicateArray]);
    const time2New = benchmark(optimized.findDuplicates, [testData.duplicateArray]);
    const improvement2 = ((time2Old - time2New) / time2Old) * 100;
    console.log(`   Inefficient: ${time2Old.toFixed(4)} ms`);
    console.log(`   Optimized:   ${time2New.toFixed(4)} ms`);
    console.log(`   Improvement: ${improvement2.toFixed(1)}% faster`);
    console.log();
    
    // Benchmark 3: Build HTML
    console.log('3. Build HTML String');
    console.log('-'.repeat(80));
    const time3Old = benchmark(inefficient.buildHTML, [testData.data]);
    const time3New = benchmark(optimized.buildHTML, [testData.data]);
    const improvement3 = ((time3Old - time3New) / time3Old) * 100;
    console.log(`   Inefficient: ${time3Old.toFixed(4)} ms`);
    console.log(`   Optimized:   ${time3New.toFixed(4)} ms`);
    console.log(`   Improvement: ${improvement3.toFixed(1)}% faster`);
    console.log();
    
    // Benchmark 4: Filter and count
    console.log('4. Filter and Count');
    console.log('-'.repeat(80));
    const time4Old = benchmark(inefficient.filterAndCount, [testData.numbers, 500]);
    const time4New = benchmark(optimized.filterAndCount, [testData.numbers, 500]);
    const improvement4 = ((time4Old - time4New) / time4Old) * 100;
    console.log(`   Inefficient: ${time4Old.toFixed(4)} ms`);
    console.log(`   Optimized:   ${time4New.toFixed(4)} ms`);
    console.log(`   Improvement: ${improvement4.toFixed(1)}% faster`);
    console.log();
    
    // Benchmark 5: Clone objects
    console.log('5. Clone Objects');
    console.log('-'.repeat(80));
    const time5Old = benchmark(inefficient.cloneObjects, [testData.items.slice(0, 100)]);
    const time5New = benchmark(optimized.cloneObjects, [testData.items.slice(0, 100)]);
    const improvement5 = ((time5Old - time5New) / time5Old) * 100;
    console.log(`   Inefficient: ${time5Old.toFixed(4)} ms`);
    console.log(`   Optimized:   ${time5New.toFixed(4)} ms`);
    console.log(`   Improvement: ${improvement5.toFixed(1)}% faster`);
    console.log();
    
    // Benchmark 6: Cache operations
    console.log('6. Cache Operations (1000 get/set operations)');
    console.log('-'.repeat(80));
    
    function testCache(CacheClass) {
        const cache = new CacheClass();
        for (let i = 0; i < 1000; i++) {
            cache.set(`key${i}`, `value${i}`);
        }
        for (let i = 0; i < 1000; i++) {
            cache.get(`key${i % 1000}`);
        }
    }
    
    const time6Old = benchmark(testCache, [inefficient.SimpleCache], 100);
    const time6New = benchmark(testCache, [optimized.OptimizedCache], 100);
    const improvement6 = ((time6Old - time6New) / time6Old) * 100;
    console.log(`   Inefficient: ${time6Old.toFixed(4)} ms`);
    console.log(`   Optimized:   ${time6New.toFixed(4)} ms`);
    console.log(`   Improvement: ${improvement6.toFixed(1)}% faster`);
    console.log();
    
    // Benchmark 7: Transform data
    console.log('7. Transform Data (map + filter + reduce)');
    console.log('-'.repeat(80));
    const time7Old = benchmark(inefficient.transformData, [testData.data]);
    const time7New = benchmark(optimized.transformData, [testData.data]);
    const improvement7 = ((time7Old - time7New) / time7Old) * 100;
    console.log(`   Inefficient: ${time7Old.toFixed(4)} ms`);
    console.log(`   Optimized:   ${time7New.toFixed(4)} ms`);
    console.log(`   Improvement: ${improvement7.toFixed(1)}% faster`);
    console.log();
    
    // Benchmark 8: Validate emails
    console.log('8. Validate Emails');
    console.log('-'.repeat(80));
    const time8Old = benchmark(inefficient.validateEmails, [testData.emails]);
    const time8New = benchmark(optimized.validateEmails, [testData.emails]);
    const improvement8 = ((time8Old - time8New) / time8Old) * 100;
    console.log(`   Inefficient: ${time8Old.toFixed(4)} ms`);
    console.log(`   Optimized:   ${time8New.toFixed(4)} ms`);
    console.log(`   Improvement: ${improvement8.toFixed(1)}% faster`);
    console.log();
    
    console.log('='.repeat(80));
    console.log('SUMMARY');
    console.log('='.repeat(80));
    console.log('All optimizations show measurable performance improvements.');
    console.log('Key optimization techniques used:');
    console.log('  - Array method chaining instead of multiple loops');
    console.log('  - Map/Set for O(1) lookups instead of arrays');
    console.log('  - Template literals and join for string building');
    console.log('  - Single-pass algorithms with reduce');
    console.log('  - Regex reuse instead of recreation');
    console.log('  - Event delegation instead of multiple listeners');
    console.log('  - Promise.all for parallel async operations');
    console.log('='.repeat(80));
}

// Run benchmarks
runBenchmarks();
